### ISSUE TITLE
Support pasting clipboard from host

### ISSUE THREAD

**Issue by fkorotkov** [cirruslabs/tart#14]
It seems it's not supported at the moment with the current API according to [this comment](https://github.com/KhaosT/MacVM/issues/14) and based on my research at the time. But let's create a tracking issue.

**Comment by fkorotkov** [comment]
It seems it will be possible. At least according to a [new Linux example](https://developer.apple.com/documentation/virtualization/running_gui_linux_in_a_virtual_machine_on_a_mac) (see "Enable copy-and-paste support between the host and the guest" section):

```swift
private func createSpiceAgentConsoleDeviceConfiguration() -> VZVirtioConsoleDeviceConfiguration {
    let consoleDevice = VZVirtioConsoleDeviceConfiguration()

    let spiceAgentPort = VZVirtioConsolePortConfiguration()
    spiceAgentPort.name = VZSpiceAgentPortAttachment.spiceAgentPortName
    spiceAgentPort.attachment = VZSpiceAgentPortAttachment()
    consoleDevice.ports[0] = spiceAgentPort

    return consoleDevice
}
```

Need to verify that it will work in macOS though.

**Comment by fkorotkov** [comment]
I've tried the above example and a bit different one:

```swift
let spiceAgentPortAttachment = VZSpiceAgentPortAttachment()
spiceAgentPortAttachment.sharesClipboard = true
let serialPortConfiguration = VZVirtioConsoleDeviceSerialPortConfiguration()
serialPortConfiguration.attachment = spiceAgentPortAttachment
configuration.serialPorts = [serialPortConfiguration]
```

with the latest Ventura and it seems not be working. It might be indicating that it only works for Linux. Will need to investigate more.

**Comment by bootstraponline** [comment]
@fkorotkov Hey, https://github.com/cirruslabs/tart/commit/048a5506df871aaddc412ddfabe2e52bec850e57 doesn't support clipboard sharing on Linux. A spice agent is necessary, similar to what you posted in https://github.com/cirruslabs/tart/issues/14#issuecomment-1147850768

https://github.com/cirruslabs/tart/pull/287 also fixes this issue. Would you prefer a standalone PR for clipboard sharing?

**Comment by fkorotkov** [comment]
Right. https://github.com/cirruslabs/tart/commit/048a5506df871aaddc412ddfabe2e52bec850e57 was only about trackpad support. @edigaryev had some security concerns as well about enabling clipboard sharing by default. Will be probably better to put the clipboard support under a flag or maybe do similar to what "Screen Sharing" app is doing e.g. only share things copied to clipboard after launching a session. It seems "Screen Sharing" either cleans the clipboard upon session creation or doing something else. Haven't investigated the behaviour in more details.

**Comment by bootstraponline** [comment]
> Right. [048a550](https://github.com/cirruslabs/tart/commit/048a5506df871aaddc412ddfabe2e52bec850e57) was only about trackpad support.

Ah, the commit message says `Support trackpad on macOS and clipboard sharing on Linux`



> @edigaryev had some security concerns as well about enabling clipboard sharing by default. Will be probably better to put the clipboard support under a flag or maybe do similar to what "Screen Sharing" app is doing e.g. only share things copied to clipboard after launching a session.

I think a flag makes sense. Some users may want an isolated VM with no clipboard support.

Does `tart` have the concept of a VM description file? In VMWare, there's [.vmx files](https://kb.vmware.com/s/article/2057902) that contain the VM configuration.

What do you think about having a `Tart.app` with a UI that enables configuring persistent VM settings that are stored on disk? https://github.com/cirruslabs/tart/issues/243#issuecomment-1248355627 mentioned the .app might be required for certain entitlements.

Parallels has a simple UI.

<img width="537" alt="image" src="https://user-images.githubusercontent.com/1173057/198894937-2debdfc2-8d95-4f65-86ef-ee2f7b29921b.png">

The settings could also be available via flag for the command line use case.

**Comment by fkorotkov** [comment]
Right now config is [stored in a JSON file](https://github.com/cirruslabs/tart/blob/89301d114efb629c70b9d3d1863b1b4251a6482b/Sources/tart/VMDirectory.swift#L14) and can be altered via `tart set` command. We were thinking about a UI similar to your screenshot so you can open `Tart.App` or run `tart run` which will display a list of available VMs with an option to change configs. We just have too many other things on our plates to find time for that.

**Comment by bootstraponline** [comment]
> We were thinking about a UI similar to your screenshot so you can open `Tart.App` or run `tart run` which will display a list of available VMs with an option to change configs. We just have too many other things on our plates to find time for that.

I looked into this and it seems like Swift Package Manager doesn't play nice with macOS apps. For full SwiftUI support (including view previews), Xcode is expecting an Xcode project.

**Comment by ThreatLockerAlex** [comment]
> with the latest Ventura and it seems not be working. It might be indicating that it only works for Linux. Will need to investigate more.

hey @fkorotkov any breakthroughs here? Just came across this thread trying to help another team at work. Not super familiar with swift but I have modified the VMInstance file of MacVM. There biggest pain point right now is clipboard sharing, hoping to either resolve let them know it is still not currently possible.

**Comment by ThreatLockerAlex** [comment]
Thank you for your efforts @fkorotkov

**Comment by fkorotkov** [comment]
All thanks to @edigaryev for double checking! 🙌

**Comment by joshuaquek** [comment]
There is another tool like `tart` that allows MacOS virtualisation as well as copy and pasting - https://github.com/insidegui/VirtualBuddy/pull/144

**Comment by lionB777** [comment]
I've found that uniclip helps quite a lot, there's still the hassle of copying and pasting files, but file sharing via smb is a good solution.

**Comment by DavidBraunDefense** [comment]
[UTM support this too.](https://docs.getutm.app/guest-support/macos/)

**Comment by Fogh** [comment]
> [All Cirrus Labs images for Sonoma and Sequoia](https://github.com/orgs/cirruslabs/packages?tab=packages&q=macos-s) got rebuild with the clipboard support. Please use upgrade to `2.26.0` version of Tart too.

I'm running `2.26.1` with latest base image, but clipboard is not shared. Do I need to run with any arguments?

**Comment by edigaryev** [comment]
@Fogh thanks for bringing this to our attention!

We've just [released a hot-fix](https://github.com/cirruslabs/macos-image-templates/pull/234) and if you re-pull the latest `ghcr.io/cirruslabs/macos-sonoma-base:latest` or `ghcr.io/cirruslabs/macos-sequoia-base:latest` it should work.

**Comment by Fogh** [comment]
> [@Fogh](https://github.com/Fogh) thanks for bringing this to our attention!
> 
> We've just [released a hot-fix](https://github.com/cirruslabs/macos-image-templates/pull/234) and if you re-pull the latest `ghcr.io/cirruslabs/macos-sonoma-base:latest` or `ghcr.io/cirruslabs/macos-sequoia-base:latest` it should work.

Sounds good! I'll try it out

**Comment by balazserdeszlogmein** [comment]
Is this available in Tahoe beta image? Don't seem to work for me.

**Comment by balazserdeszlogmein** [comment]
> Which exact image are you using?

I'm using `ghcr.io/cirruslabs/macos-tahoe-vanilla:26.0`.

**Comment by balazserdeszlogmein** [comment]
Sadly that don't seem to work for me.

- `brew install tart-guest-agent` does nothing on the tart vm, `brew install cirruslabs/cli/tart-guest-agent` works though.
- but even after I copy the plist file, there is no copy-paste feature.

**Comment by balazserdeszlogmein** [comment]
@edigaryev 

- using 2.25.0
- `tart run [vm name] --dir=share:[path/to/a/folder]:ro`
- I am not aware of what that is, so I would assume no.

**Comment by balazserdeszlogmein** [comment]
I did two things, one of those fixed it
- upgraded tart to 2.28.0 on host
- installed tart on guest (not just tart-guest-agent)

now it works. Thanks!
